app/code/Bulbulatory/Recommendations/Setup/Patch/Schema/UpgradeHashRecommendations.php

```diff
@@ -0,0 +1,50 @@
+<?php
+namespace Bulbulatory\Recommendations\Setup\Patch\Schema;
+
+use Magento\Framework\Setup\Patch\SchemaPatchInterface;
+use Magento\Framework\Setup\ModuleDataSetupInterface;
+use Magento\Framework\DB\Ddl\TriggerFactory;
+use Magento\Framework\DB\Ddl\Trigger;
+
+class UpgradeHashRecommendations implements SchemaPatchInterface
+{
+    private $moduleDataSetup;
+    private $triggerFactory;
+
+    public function __construct(
+        ModuleDataSetupInterface $moduleDataSetup,
+        TriggerFactory $triggerFactory
+    ) {
+        $this->moduleDataSetup = $moduleDataSetup;
+        $this->triggerFactory = $triggerFactory;
+    }
+
+    public static function getDependencies()
+    {
+        return [];
+    }
+
+    public function getAliases()
+    {
+        return [];
+    }
+
+    public function apply()
+    {
+        $setup = $this->moduleDataSetup;
+        $setup->startSetup();
+
+        $trigger = $this->triggerFactory->create()
+            ->setName('trg_bulbulatory_recommendations_after_insert')
+            ->setTime(Trigger::TIME_BEFORE)
+            ->setEvent('INSERT')
+            ->setTable($setup->getTable('bulbulatory_recommendations'));

```

Comment:
Using triggers is not a good idea for generating hash. It's better to generate it in the code because when someone new enters the project, he or she will take a look at the code and will know straight away what it does. With triggers it's not so obvious. Debugging code that uses triggers is also not as easy as debugging just the code, it may be very difficult to track the potential issue. It's best to generate hash in the code, just before saving recommendation data to the database (will be done in a later task).

Reply:
@TM18

> Using triggers is not a good idea for generating hash. It's better to generate it in the code because when someone new enters the project, he or she will take a look at the code and will know straight away what it does. With triggers it's not so obvious. Debugging code that uses triggers is also not as easy as debugging just the code, it may be very difficult to track the potential issue. It's best to generate hash in the code, just before saving recommendation data to the database (will be done in a later task).

I misunderstood this field description. I thought that _hash.bulbulatory_recommendations_ should automatically generate when a new row is created in table _bulbulatory_recommendations_.

I deleted file _**Bulbulatory/Recommendations/Setup/Patch/Schema/UpgradeHashRecommendations.php**_